sphinx/registry.py

```diff
@@ -153,11 +153,11 @@ def preload_builder(self, app: "Sphinx", name: str) -> None:
 
             self.load_extension(app, entry_point.module)
 
-    def create_builder(self, app: "Sphinx", name: str) -> Builder:
+    def create_builder(self, app: "Sphinx", name: str, env: BuildEnvironment) -> Builder:
         if name not in self.builders:
             raise SphinxError(__('Builder name %s not registered') % name)
 
-        return self.builders[name](app)
+        return self.builders[name](app, env)
```

Comment:
This is a breaking change if the 3rd party builder overrides the `__init__()` method. We need to handle `TypeError` if it does not take the `env` argument:
```suggestion
        try:
            return self.builders[name](app, env)
        except TypeError:
            warnings.warn("The 'env' argument to Builder will be required from Sphinx 7.")
            builder = self.builders[name](app)
            builder.set_environment(env)
            return builder
```

Reply:
As I added the warning code to `Builder.__init__`, I think just making the default argument `None` would preserve backwards compatability here?

A

Reply:
Does this builder work?

```
class MyBuilder(Builder):
    def __init__(self, app):
        # my initialization code here
        ...

        super().__init__(app)
```

Reply:
Yep, note in `builders/__init__.py`:

```python
class Builder:
    ... # class setup elided

    def __init__(self, app: "Sphinx", env: BuildEnvironment = None) -> None:
        # instance setup elided

        if env is not None:
            self.env: BuildEnvironment = env
            self.env.set_versioning_method(self.versioning_method,
                                           self.versioning_compare)
        else:
            warnings.warn("The 'env' argument to Builder will be required from Sphinx 7.",
                          RemovedInSphinx70Warning, stacklevel=2)
```

So `env` will default to `None` and then raise a warning.

A

Reply:
Please try this small custom builder.

On my local, it causes a crash:
```
$ sphinx-build -b custom . _build/custom
```
```
Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/sphinx/cmd/build.py", line 272, in build_main
    app = Sphinx(args.sourcedir, args.confdir, args.outputdir,
  File "/Users/tkomiya/work/sphinx/sphinx/application.py", line 253, in __init__
    self.builder = self.create_builder(buildername)
  File "/Users/tkomiya/work/sphinx/sphinx/application.py", line 326, in create_builder
    return self.registry.create_builder(self, name, self.env)
  File "/Users/tkomiya/work/sphinx/sphinx/registry.py", line 161, in create_builder
    return self.builders[name](app, env)
TypeError: __init__() takes 2 positional arguments but 3 were given
```

Reply:
[This https://github.com/tk0miya/custom_builder](https://github.com/tk0miya/custom_builder)

Reply:
Ahh, yes I see, sorry. One second.

A

Reply:
@tk0miya this should be fixed with the latest commit.

A